components/dashboard/src/service/json-rpc-envvar-client.ts

```diff
@@ -0,0 +1,229 @@
+
+        const variable: UserEnvVarValue = {
+            name: req.name,
+            value: req.value,
+            repositoryPattern: req.repositoryPattern,
+        };
+        variable.repositoryPattern = UserEnvVar.normalizeRepoPattern(variable.repositoryPattern);
+
+        await getGitpodService().server.setEnvVar(variable);
+
+        const updatedUserEnvVars = await getGitpodService().server.getAllEnvVars();
+        const updatedUserEnvVar = updatedUserEnvVars.find(
+            (v) => v.name === variable.name && v.repositoryPattern === variable.repositoryPattern,
+        );
+        if (!updatedUserEnvVar) {
+            throw new ConnectError("could not update env variable", Code.Internal);
+        }
+
+        response.environmentVariable = converter.toUserEnvironmentVariable(updatedUserEnvVar);
+
+        return response;
+    }
+
+    async deleteUserEnvironmentVariable(
+        req: PartialMessage<DeleteUserEnvironmentVariableRequest>,
+    ): Promise<DeleteUserEnvironmentVariableResponse> {
+        if (!req.envVarId) {
+            throw new ConnectError("invalid argument", Code.InvalidArgument);
+        }
+
+        const variable: UserEnvVarValue = {
+            id: req.envVarId,
+            name: "",
+            value: "",
+            repositoryPattern: "",
+        };
+
+        await getGitpodService().server.deleteEnvVar(variable);
+
+        const response = new DeleteUserEnvironmentVariableResponse();
+        return response;
+    }
+
+    async listConfigurationEnvironmentVariables(
+        req: PartialMessage<ListConfigurationEnvironmentVariablesRequest>,
+    ): Promise<ListConfigurationEnvironmentVariablesResponse> {
+        if (!req.configurationId) {
+            throw new ConnectError("configurationId is not set", Code.InvalidArgument);
+        }
+
+        const result = new ListConfigurationEnvironmentVariablesResponse();
+        const projectEnvVars = await getGitpodService().server.getProjectEnvironmentVariables(req.configurationId);
+        result.environmentVariables = projectEnvVars.map((i) => converter.toConfigurationEnvironmentVariable(i));
+
+        return result;
+    }
+
+    async updateConfigurationEnvironmentVariable(
+        req: PartialMessage<UpdateConfigurationEnvironmentVariableRequest>,
+    ): Promise<UpdateConfigurationEnvironmentVariableResponse> {
+        if (!req.envVarId) {
+            throw new ConnectError("envVarId is not set", Code.InvalidArgument);
+        }
+        if (!req.configurationId) {
+            throw new ConnectError("configurationId is not set", Code.InvalidArgument);
+        }
+
+        const response = new UpdateConfigurationEnvironmentVariableResponse();
+
+        const projectEnvVars = await getGitpodService().server.getProjectEnvironmentVariables(req.configurationId);
+        const projectEnvVarfound = projectEnvVars.find((i) => i.id === req.envVarId);
+        if (projectEnvVarfound) {
```

Comment:
Server has guard for this case, that if not found then create. So is it safe to update without check in shim? Or we want to make sure that it only do `update` but not `upsert`?

Reply:
update should only update